engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -184,6 +192,19 @@ private void decrementParallelGatewaySequenceFlow(
     elementInstanceState.updateInstance(flowScopeInstance);
   }
 
+  private void decrementInclusiveGatewayGatewaySequenceFlow(
+      final ProcessInstanceRecord value, final ElementInstance flowScopeInstance) {
+    // Inclusive gateways can have more than one incoming sequence flow, we need to decrement the
+    // active sequence flows based on the incoming count.
+
+    final var executableFlowNode =
+        processState.getFlowElement(
+            value.getProcessDefinitionKey(), value.getElementIdBuffer(), ExecutableFlowNode.class);
+    final var size = executableFlowNode.getIncoming().size();
+    IntStream.range(0, size).forEach(i -> flowScopeInstance.decrementActiveSequenceFlows());
```

Comment:
❌ This only works correctly if all incoming sequence flows were taken. That is always the case for parallel gateway, but not for the inclusive gateway because it can be that there weren't enough tokens. For example, if just 2 of 3 incoming sequence flows were taken and there are no other tokens left, then the inclusive gateway is activated.

Therefore, we should only decrement the flowscope instance's number of active sequence flows with the number that was actually taken to activate this gateway.

💡 I think we have that data available. On line 92, you decrement the number of taken sequence flows (because the gateway is now activating). This lowers each value by 1. We could simply have this `decrementNumberOfTakenSequenceFlows` method return the number of sequence flows that were taken, store that value in a variable and pass it to the `decrementInclusiveGatewayGatewaySequenceFlow` method. 

Please let me know if you have any questions about this.

Reply:
> if just 2 of 3 incoming sequence flows were taken and there are no other tokens left, then the inclusive gateway is activated

Hi, @korthout  check the manual test, it's work fine,it's been handled by "_InclusiveGatewayProcessor.takeSequenceFlow_"

![20220715015907](https://user-images.githubusercontent.com/26370117/179183115-cd975356-c094-4e47-b42a-56a82c9b3937.png)

![20220715020012](https://user-images.githubusercontent.com/26370117/179183144-644f3e1a-9560-48d6-a29b-8fd1df916ff8.png)

Reply:
Yes that process seems to work correctly, but the number of active sequence flows that are decremented in that method call is lowered too many times. In more complex processes this would become clear.

In an attempt to find such a process, I noticed that the issue is actually even bigger. I'll point it out at the respective code location in another comment. Let's first focus on that, and then we can come back here.